modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m

```diff
@@ -1066,6 +1063,9 @@ - (void)startDrag:(NSDragOperation)operation withItems:(NSArray<NSDraggingItem*>
                 if (offset.y > imageHalfY || offset.y < -imageHalfY) {
                     offset.y = imageHalfY * (offset.y > 0 ? 1 : -1);
                 }
+
+                dragPoint.x -= offset.x;
+                dragPoint.y -= offset.y;
```

Comment:
This works fine now, _if_ the offset values are smaller than the imageHalf X,Y values, which doesn't make sense anymore:

- If you drag from {0,0} (top-left), it works from that position
- But if you drag from {imageW, imageH} (bottom-right), the offset is clamped to the image centre, which it is not expected.

So now we need to change the above lines to clamp the offset between 0,0 and imageW,H (though Windows doesn't clamp the offset at all).

Reply:
Out of curiosity I removed offset clamping (macOS snaps the drag image regardless) and technically it still works, but when you cancel the drag the image snaps back to the offset position. With high enough (or negative) values drag image can animate back even outside the application bounds, which makes little sense and looks weird. I'll clamp it between 0,0 and imageW,H as you mentioned.

I also didn't find a way to disable above drag image "snapping". However, if it makes sense to clamp it on macOS, maybe it would also make sense to introduce similar limits to offsets on other platforms (probably as a separate issue)?

Reply:
Right, there is a built-in native animation from the offset point to the mouse cursor (in and out). 

When the drag event is cancelled, the image returns to the offset location with this animation. We still want that, so I wouldn't remove the animation (if we could), but of course, with the original position of the source, no far away from it, so it makes sense to clamp between the image bounds. 

And doing the same (clamping) for Windows would be a separate issue, indeed.

Reply:
By the way, this is how you can disable the cancel animation:

```
diff --git a/modules/javafx.graphics/src/main/native-glass/mac/GlassDraggingSource.m b/modules/javafx.graphics/src/main/native-glass/mac/GlassDraggingSource.m
index 50a79cabf6..373537f24e 100644
--- a/modules/javafx.graphics/src/main/native-glass/mac/GlassDraggingSource.m
+++ b/modules/javafx.graphics/src/main/native-glass/mac/GlassDraggingSource.m
@@ -40,6 +40,7 @@
 
 - (NSDragOperation)draggingSession:(NSDraggingSession *)session sourceOperationMaskForDraggingContext:(NSDraggingContext)context
 {
+     session.animatesToStartingPositionsOnCancelOrFail = NO;
     // The Command key masks out every operation other than NSDragOperationGeneric. We want
     // internal Move events to get through this filtering so we copy the Move bit into the
     // Generic bit and treat Generic as a synonym for Move.
```

I'm still in favour of this native animation, as opposed to the Windows case that doesn't have it. This can be a separated issue as well (unifying the animation in all platforms).

Reply:
I wanted to chime in, thought it might be unrelated:

In the context of a docking framework, it is impossible to disable the cancel animation.  We should have a way to disable it.  The use case is when the user drags and drops a tab/pane outside of the dockable windows.  In this case a new window should appear at the drop location, without the drag image animated back to the start position.

Another thing that is impossible is to control the drag image: sometimes one needs a scaled icon, sometimes one wants the full content.

These two issues basically prevent implementing a docking framework via DnD, one needs to use other solution, for example
https://github.com/andy-goryachev/FxDock

Reply:
I'm not sure if I follow, but about the cancel animation, do you mean that is not possible to disable it on macOS? Have you tested `session.animatesToStartingPositionsOnCancelOrFail`? 
As for the image, `DragBoard::setDragView` allows you setting any image that you need for the drag gesture, or am I missing something?

Reply:
it's been a while - java8 time frame - back then I was unable to find a way to disable animation on cancel.

as for image - I was setting large image and fx tried to scale it to be smaller, I was unable to force it to show the unscaled full size image.

also, why does the drag image look fuzzy on mac retina?